<a href="https://colab.research.google.com/github/lamyse1/Data-Engineering-Projects/blob/main/DE_Graded_Project_Week2_lamyse_Ammar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Working Towards The DE Project: My Dashboard**

# **1. Set Up the Sales Dataset:**

In [21]:
import pandas as pd

csv_url = "https://raw.githubusercontent.com/lamyse1/Data-Engineering-Projects/main/D.E.%20Graded%20Project.%20Store%20Sales%20Data.csv"
sales_data = pd.read_csv(csv_url)

# Preview the first few rows
print(sales_data.head())


       date  product ID  sales amount   store location 
0  1/1/2013           1             13                1
1  1/2/2013           1             11                1
2  1/3/2013           1             14                1
3  1/4/2013           1             13                1
4  1/5/2013           1             10                1


In [22]:
# Check column names and first few rows
print("Column Names:", sales_data.columns)
print("\nFirst 5 Rows:")
print(sales_data.head())

# Check for missing values
print("\nMissing Values Per Column:")
print(sales_data.isnull().sum())

# Get summary statistics for numerical columns
print("\nBasic Statistics:")
print(sales_data.describe())

# Check data types
print("\nData Types:")
print(sales_data.dtypes)


Column Names: Index(['date', 'product ID', 'sales amount ', 'store location '], dtype='object')

First 5 Rows:
       date  product ID  sales amount   store location 
0  1/1/2013           1             13                1
1  1/2/2013           1             11                1
2  1/3/2013           1             14                1
3  1/4/2013           1             13                1
4  1/5/2013           1             10                1

Missing Values Per Column:
date               0
product ID         0
sales amount       0
store location     0
dtype: int64

Basic Statistics:
          product ID  sales amount   store location 
count  913000.000000  913000.000000    913000.000000
mean       25.500000      52.250287         5.500000
std        14.430878      28.801144         2.872283
min         1.000000       0.000000         1.000000
25%        13.000000      30.000000         3.000000
50%        25.500000      47.000000         5.500000
75%        38.000000      70.000000   

In [23]:
# remove any extra spaces from column names
sales_data.columns = sales_data.columns.str.strip()

# Verify column names after cleaning
print("Updated Column Names:", sales_data.columns)


Updated Column Names: Index(['date', 'product ID', 'sales amount', 'store location'], dtype='object')


In [24]:
# Define mapping for store numbers
store_mapping = {
    1: "New York",
    2: "Los Angeles",
    3: "Chicago",
    4: "Houston",
    5: "Phoenix",
    6: "Philadelphia",
    7: "San Antonio",
    8: "San Diego",
    9: "Dallas",
    10: "San Jose"
}

# Apply mapping to store_location column
sales_data["store location"] = sales_data["store location"].map(store_mapping)

# Confirm mapping applied correctly
print("\nMapped Store Locations:")
print(sales_data[["store location"]].drop_duplicates())  # Display unique locations to verify



Mapped Store Locations:
      store location
0           New York
1826     Los Angeles
3652         Chicago
5478         Houston
7304         Phoenix
9130    Philadelphia
10956    San Antonio
12782      San Diego
14608         Dallas
16434       San Jose


# **Perform Basic Sales Analysis (Grouping & Aggregations)**

In [25]:
# 1. Calculate total sales across all locations
total_sales = sales_data["sales amount"].sum()
print(f"\nTotal Sales (All Locations): {total_sales}")


Total Sales (All Locations): 47704512


In [26]:
# 2. Sales by store location
sales_by_location = sales_data.groupby("store location")["sales amount"].sum().reset_index()
print("\nTotal Sales by Store Location:")
print(sales_by_location)


Total Sales by Store Location:
  store location  sales amount
0        Chicago       5435144
1         Dallas       5025976
2        Houston       5012639
3    Los Angeles       6120128
4       New York       4315603
5   Philadelphia       3627670
6        Phoenix       3631016
7    San Antonio       3320009
8      San Diego       5856169
9       San Jose       5360158


In [27]:
# 3. Sales by product
sales_by_product = sales_data.groupby("product ID")["sales amount"].sum().reset_index()
print("\nTotal Sales by Product:")
print(sales_by_product)


Total Sales by Product:
    product ID  sales amount
0            1        401384
1            2       1069564
2            3        669087
3            4        401907
4            5        335230
5            6       1068281
6            7       1068777
7            8       1405108
8            9        938379
9           10       1337133
10          11       1271925
11          12       1271534
12          13       1539621
13          14       1071531
14          15       1607442
15          16        468480
16          17        602486
17          18       1538876
18          19        736892
19          20        867641
20          21        736190
21          22       1469971
22          23        534979
23          24       1205975
24          25       1473334
25          26        869981
26          27        402628
27          28       1604713
28          29       1271240
29          30        736554
30          31       1070845
31          32        803107
32          33    

In [28]:
# 4. Find the best-selling store location
best_selling_store = sales_by_location.loc[sales_by_location["sales amount"].idxmax()]

print("\nBest-Selling Store Location:")
print(f"{best_selling_store['store location']} with total sales of {best_selling_store['sales amount']}")



Best-Selling Store Location:
Los Angeles with total sales of 6120128


In [29]:
# Group sales by product ID
sales_by_product = sales_data.groupby("product ID")["sales amount"].sum().reset_index()

# 5. Find the best-selling product
best_selling_product = sales_by_product.loc[sales_by_product["sales amount"].idxmax()]

print("\nBest-Selling Product:")
print(f"Product {best_selling_product['product ID']} with total sales of {best_selling_product['sales amount']}")



Best-Selling Product:
Product 15 with total sales of 1607442


In [30]:
# 6. Summary statistics for sales amount
print("\nSales Amount Statistics:")
print(sales_data["sales amount"].describe())


Sales Amount Statistics:
count    913000.000000
mean         52.250287
std          28.801144
min           0.000000
25%          30.000000
50%          47.000000
75%          70.000000
max         231.000000
Name: sales amount, dtype: float64


# **2. Create a Basic Visualization with Dash:**

In [31]:
!pip install dash
!pip install jupyter-dash
!pip install plotly


In [32]:
import dash
from dash import dcc, html
import plotly.express as px
from jupyter_dash import JupyterDash
import pandas as pd
from dash.dependencies import Input, Output

In [33]:
# Aggregate sales data by store location
sales_by_location = sales_data.groupby("store location")["sales amount"].sum().reset_index()


In [34]:
# Create a bar chart for sales by store location ( following the example on the exercise requirements sheet)
fig = px.bar(
    sales_by_location,
    x="store location",
    y="sales amount",
    title="Sales by Store Location"
)


In [35]:

# Initialize the Dash app
app = dash.Dash(__name__)

# Add "All Locations" option for store sales
all_locations = pd.DataFrame({"store location": ["All Locations"], "sales amount": [sales_by_location["sales amount"].sum()]})
sales_by_location_with_all = pd.concat([all_locations, sales_by_location], ignore_index=True)

#  Add "All Products" option for product sales
all_products = pd.DataFrame({"product ID": ["All Products"], "sales amount": [sales_by_product["sales amount"].sum()]})
sales_by_product_with_all = pd.concat([all_products, sales_by_product], ignore_index=True)

# Define the app layout with two separate sections
app.layout = html.Div(children=[
    html.H1(children="Sales Dashboard"),

    # Section 1: Store Location Sales
    html.H3(children="Sales by Store Location"),
    dcc.Dropdown(
        id="store-selector",
        options=[{"label": loc, "value": loc} for loc in sales_by_location_with_all["store location"]],
        value="All Locations",
        clearable=False
    ),
    dcc.Graph(id="sales-location-chart"),

    # Section 2: Product Sales
    html.H3(children="Sales by Product"),
    dcc.Dropdown(
        id="product-selector",
        options=[{"label": str(prod), "value": prod} for prod in sales_by_product_with_all["product ID"]],
        value="All Products",
        clearable=False
    ),
    dcc.Graph(id="sales-product-chart")
])

#  Define Callback for Store Sales Chart
@app.callback(
    Output("sales-location-chart", "figure"),
    [Input("store-selector", "value")]
)
def update_store_chart(selected_store):
    df = sales_by_location_with_all if selected_store == "All Locations" else \
         sales_by_location[sales_by_location["store location"] == selected_store]

    fig = px.bar(
        df,
        x="store location",
        y="sales amount",
        title=f"Sales for {selected_store}"
    )
    return fig

#  Define Callback for Product Sales Chart
@app.callback(
    Output("sales-product-chart", "figure"),
    [Input("product-selector", "value")]
)
def update_product_chart(selected_product):
    df = sales_by_product_with_all if selected_product == "All Products" else \
         sales_by_product[sales_by_product["product ID"] == selected_product]

    fig = px.bar(
        df,
        x="product ID",
        y="sales amount",
        title=f"Sales for {selected_product}"
    )
    return fig

# Run the Dash server
if __name__ == '__main__':
    app.run_server(debug=True)


<IPython.core.display.Javascript object>

In [36]:
print("\nBest-Selling Store Location:")
print(f"{best_selling_store['store location']} with total sales of {best_selling_store['sales amount']}")

print("\nBest-Selling Product:")
print(f"Product {best_selling_product['product ID']} with total sales of {best_selling_product['sales amount']}")



Best-Selling Store Location:
Los Angeles with total sales of 6120128

Best-Selling Product:
Product 15 with total sales of 1607442


# **Sales Overview:**

The dataset contains sales records for multiple stores and products.
The total number of transactions is 913,080.
The highest single transaction amount recorded is 231.

 Best-Selling Store Location: Los Angeles has the highest total sales (6,120,128).
 Best-Selling Product: Product 15 has the highest total sales (1,607,442).

 Trends & Patterns from Dash Visualization:

Some locations generate more sales than others, showing potential for market expansion like Sandiego, chicago and The dropdown filtering allows quick insights into sales trends for specific stores/products.

# **WEEK 2 Exercise:**

# **Set Up Mongo DB **

In [37]:
!pip install pymongo


In [38]:
import pymongo


MONGO_URI = "mongodb+srv://lamyseammar:Laura9966@cluster0.pfzed.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"

# Connect to MongoDB
client = pymongo.MongoClient(MONGO_URI)

# Create the database and collection
db = client["sales_db"]  # Database: sales_db
sales_collection = db["sales"]  # Collection: sales

# Confirm connection
print("✅ Connected to MongoDB Atlas and created 'sales_db' with 'sales' collection.")



✅ Connected to MongoDB Atlas and created 'sales_db' with 'sales' collection.


In [39]:
from google.colab import files

uploaded = files.upload()


Saving D.E. Graded Project. Store Sales Data.csv to D.E. Graded Project. Store Sales Data (1).csv


In [40]:
import pandas as pd


csv_filename = "D.E. Graded Project. Store Sales Data (1).csv"

# Load the CSV file into a Pandas DataFrame
sales_data = pd.read_csv(csv_filename)

# Display the first few rows to verify
print("✅ First 5 rows of the dataset:")
print(sales_data.head())


✅ First 5 rows of the dataset:
       date  product ID  sales amount   store location 
0  1/1/2013           1             13                1
1  1/2/2013           1             11                1
2  1/3/2013           1             14                1
3  1/4/2013           1             13                1
4  1/5/2013           1             10                1


In [41]:
# Convert DataFrame rows to a dictionary format for MongoDB
sales_records = sales_data.to_dict(orient="records")

# Insert the sales data into MongoDB
sales_collection.insert_many(sales_records)

print(f"✅ Successfully inserted {len(sales_records)} records into MongoDB.")


✅ Successfully inserted 913000 records into MongoDB.


In [42]:
# Count number of records in MongoDB
count = sales_collection.count_documents({})
print(f"✅ Total records in MongoDB: {count}")

# Fetch and display the first 5 records from MongoDB
print("\n🔹 First 5 records from MongoDB:")
for record in sales_collection.find().limit(5):
    print(record)


✅ Total records in MongoDB: 913000

🔹 First 5 records from MongoDB:
{'_id': ObjectId('67a07e0fcd78498ad38e9302'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 13, 'store location ': 1}
{'_id': ObjectId('67a07e0fcd78498ad38e9303'), 'date': '1/2/2013', 'product ID': 1, 'sales amount ': 11, 'store location ': 1}
{'_id': ObjectId('67a07e0fcd78498ad38e9304'), 'date': '1/3/2013', 'product ID': 1, 'sales amount ': 14, 'store location ': 1}
{'_id': ObjectId('67a07e0fcd78498ad38e9305'), 'date': '1/4/2013', 'product ID': 1, 'sales amount ': 13, 'store location ': 1}
{'_id': ObjectId('67a07e0fcd78498ad38e9306'), 'date': '1/5/2013', 'product ID': 1, 'sales amount ': 10, 'store location ': 1}


# **Perform CRUD Operations on MongoDB**

 Create: Insert Additional Sales Records

In [43]:
# New sales records to insert
new_sales = [
    {"date": "2/2/2025", "product ID": 51, "sales amount": 120, "store location": 3},
    {"date": "2/3/2025", "product ID": 52, "sales amount": 200, "store location": 7}
]

# Insert new records
sales_collection.insert_many(new_sales)

# Confirm insertion
print("✅ Successfully added new sales records.")


✅ Successfully added new sales records.


Read: Query Data Based on Store, Product, or Date

In [44]:
# Fetch sales by store location
store_sales = list(sales_collection.find({"store location": 3}).limit(5))
print("\n🔹 Sales from Store Location 3:")
for sale in store_sales:
    print(sale)

# Fetch sales by product ID
product_sales = list(sales_collection.find({"product ID": 51}).limit(5))
print("\n🔹 Sales for Product ID 51:")
for sale in product_sales:
    print(sale)

# Fetch sales by date
date_sales = list(sales_collection.find({"date": "1/1/2013"}).limit(5))
print("\n🔹 Sales on 1/1/2013:")
for sale in date_sales:
    print(sale)



🔹 Sales from Store Location 3:
{'_id': ObjectId('67a08392cd78498ad39c816a'), 'date': '2/2/2025', 'product ID': 51, 'sales amount': 120, 'store location': 3}

🔹 Sales for Product ID 51:
{'_id': ObjectId('67a08392cd78498ad39c816a'), 'date': '2/2/2025', 'product ID': 51, 'sales amount': 120, 'store location': 3}

🔹 Sales on 1/1/2013:
{'_id': ObjectId('67a07e0fcd78498ad38e9302'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 13, 'store location ': 1}
{'_id': ObjectId('67a07e0fcd78498ad38e9a24'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 12, 'store location ': 2}
{'_id': ObjectId('67a07e0fcd78498ad38ea146'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 19, 'store location ': 3}
{'_id': ObjectId('67a07e0fcd78498ad38ea868'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 10, 'store location ': 4}
{'_id': ObjectId('67a07e0fcd78498ad38eaf8a'), 'date': '1/1/2013', 'product ID': 1, 'sales amount ': 11, 'store location ': 5}


Update: Modify Sales Amount for a Specific Record

In [45]:
# Update the sales amount for product ID 51 on "2/2/2025"
sales_collection.update_one(
    {"date": "2/2/2025", "product ID": 51},
    {"$set": {"sales amount": 500}}
)

# Confirm update
updated_record = sales_collection.find_one({"date": "2/2/2025", "product ID": 51})
print("\n✅ Updated Sales Record:")
print(updated_record)



✅ Updated Sales Record:
{'_id': ObjectId('67a08392cd78498ad39c816a'), 'date': '2/2/2025', 'product ID': 51, 'sales amount': 500, 'store location': 3}


Delete: Remove Outdated Sales Records

In [46]:
# Delete sales records before 2015
delete_result = sales_collection.delete_many({"date": {"$lt": "1/1/2015"}})

# Confirm deletion
print(f"\n✅ Deleted {delete_result.deleted_count} outdated sales records.")



✅ Deleted 1000 outdated sales records.



# **Visualize data from MongoDB using Dash**

# Retrieve Updated Sales Data from MongoDB

In [47]:
# Fetch sales data from MongoDB and convert it to a DataFrame
mongo_data = list(sales_collection.find())

# Convert to DataFrame
sales_df = pd.DataFrame(mongo_data)

# Drop MongoDB's _id column since we dont need it  for visualization
sales_df.drop(columns=["_id"], inplace=True)

# Display first few rows
print("✅ Updated Sales Data from MongoDB:")
print(sales_df.head())


✅ Updated Sales Data from MongoDB:
       date  product ID  sales amount   store location   sales amount  \
0  1/2/2013           1           11.0              1.0           NaN   
1  1/3/2013           1           14.0              1.0           NaN   
2  1/4/2013           1           13.0              1.0           NaN   
3  1/5/2013           1           10.0              1.0           NaN   
4  1/6/2013           1           12.0              1.0           NaN   

   store location  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  


# Modify Dash App to Use MongoDB Data

In [48]:
import dash
from dash import dcc, html
import plotly.express as px
from dash.dependencies import Input, Output

# Initialize the Dash app
app = dash.Dash(__name__)

# Fetch updated sales data from MongoDB
def get_sales_data():
    mongo_data = list(sales_collection.find())
    df = pd.DataFrame(mongo_data)
    df.drop(columns=["_id"], inplace=True)
    return df

# Define the app layout
app.layout = html.Div(children=[
    html.H1(children="Sales Dashboard"),

    # Store Location Sales Chart
    html.H3(children="Sales by Store Location"),
    dcc.Graph(id="sales-location-chart"),

    # Product Sales Chart
    html.H3(children="Sales by Product"),
    dcc.Graph(id="sales-product-chart")
])

# Define callback to update charts with latest MongoDB data
@app.callback(
    [Output("sales-location-chart", "figure"),
     Output("sales-product-chart", "figure")],
    Input("sales-location-chart", "id")
)
def update_charts(_):
    df = get_sales_data()

    # Aggregate sales by store location
    sales_by_location = df.groupby("store location")["sales amount"].sum().reset_index()
    fig_location = px.bar(sales_by_location, x="store location", y="sales amount", title="Sales by Store Location")

    # Aggregate sales by product
    sales_by_product = df.groupby("product ID")["sales amount"].sum().reset_index()
    fig_product = px.bar(sales_by_product, x="product ID", y="sales amount", title="Sales by Product")

    return fig_location, fig_product

# Run the Dash server
if __name__ == '__main__':
    app.run_server(debug=True)


<IPython.core.display.Javascript object>

# **Documentation Content**

# CRUD Operations on MongoDB - Sales Data

## 1.Create (Insert New Records)
- Two new sales records were manually inserted into MongoDB.
- Example Record:
  {"date": "2/2/2025", "product ID": 51, "sales amount": 120, "store location": 3}
- This allows adding future sales data dynamically.

## 2.Read (Query Data from MongoDB)
- Queried sales based on store location, product ID, and date.
- Example Queries:
  - Find all sales from store location 3.
  - Find all sales for product ID 51.
  - Find all sales on "1/1/2013".
- This enables analysis of sales trends and key performance metrics.

## 3.Update (Modify Records)
- Sales amount for Product ID 51 on "2/2/2025" was updated from 120 to 500.
- Example Query:
  ```python
  sales_collection.update_one(
      {"date": "2/2/2025", "product ID": 51},
      {"$set": {"sales amount": 500}}
  )


## 4. Delete (Remove Outdated Records)
- All sales records before "1/1/2015" were deleted to maintain database efficiency.
- Example Query:
```python
sales_collection.delete_many({"date": {"$lt": "1/1/2015"}})

Removing old records optimizes storage and speeds up queries.



# Why CRUD Operations Are Important

CRUD (Create, Read, Update, Delete) operations ensure efficient database management:

Create → Adds new sales data dynamically.

Read → Retrieves key sales insights (best-selling stores, products).

Update → Corrects incorrect data and updates transactions.

Delete → Removes outdated or unnecessary data.

# Finally:
Sales data is stored in MongoDB and dynamically updated.
Dash visualization displays real-time sales data after CRUD operations.
The project successfully integrates a NoSQL database with interactive dashboards.

In [52]:
documentation = """
# CRUD Operations on MongoDB - Sales Data

## 1.Create (Insert New Records)
- Two new sales records were manually inserted into MongoDB.
- Example Record:
  {"date": "2/2/2025", "product ID": 51, "sales amount": 120, "store location": 3}
- This allows adding future sales data dynamically.

## 2.Read (Query Data from MongoDB)
- Queried sales based on store location, product ID, and date.
- Example Queries:
  - Find all sales from store location 3.
  - Find all sales for product ID 51.
  - Find all sales on "1/1/2013".
- This enables analysis of sales trends and key performance metrics.

## 3. Update (Modify Records)
- Sales amount for Product ID 51 on "2/2/2025" was updated from 120 to 500.
- Example Query:
  sales_collection.update_one(
      {"date": "2/2/2025", "product ID": 51},
      {"$set": {"sales amount": 500}}
  )

- Updating records ensures accuracy and up-to-date financial reporting.

## 4.Delete (Remove Outdated Records)
- All sales records before "1/1/2015" were deleted to maintain database efficiency.
- Example Query:
  sales_collection.delete_many({"date": {"$lt": "1/1/2015"}})

- Removing old records optimizes storage and speeds up queries.

## Why CRUD Operations Are Important
CRUD (Create, Read, Update, Delete) operations ensure efficient database management:
- **Create** → Adds new sales data dynamically.
- **Read** → Retrieves key sales insights (best-selling stores, products).
- **Update** → Corrects incorrect data and updates transactions.
- **Delete** → Removes outdated or unnecessary data.

## Finally
- **Sales data is stored in MongoDB** and dynamically updated.
- **Dash visualization displays real-time sales data** after CRUD operations.
- **The project successfully integrates a NoSQL database with interactive dashboards.**
"""

# Save documentation to a text file
with open("crud_documentation.txt", "w") as file:
    file.write(documentation)

print("✅ Documentation saved as 'crud_documentation.txt'. Upload it to GitHub for submission.")


✅ Documentation saved as 'crud_documentation.txt'. Upload it to GitHub for submission.
